File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CsrToCscBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Backward pass for {@link CsrToCsc}: apply the inverse value-permutation to produce
 * the gradient of the CSR values input.
 *
 * <p>C++ op name: {@code csr_to_csc_bp}
 *
 * <p><b>Background:</b> The forward {@code csr_to_csc} applies a counting-sort permutation
 * {@code P} that maps CSR entry positions to CSC entry positions.  The gradient of
 * {@code cscValues = P(csrValues)} with respect to {@code csrValues} is simply:
 * {@code dCsrValues[e] = gradCscValues[P(e)]}.  The inverse permutation {@code P⁻¹} is
 * computed device-side by re-applying {@code csr_to_csc} to the gradient arrays with the
 * matrix dimensions swapped (treating the CSC of A as the CSR of A^T).
 *
 * <p><b>Inputs (5 arrays):</b>
 * <ol>
 *   <li>{@code csrColIdx}     – 1D [nnz], INT32/INT64 — CSR column indices (forward input[1])</li>
 *   <li>{@code csrRowPtr}     – 1D [rows+1], same INT — CSR row pointers (forward input[2])</li>
 *   <li>{@code cscRowIdx}     – 1D [nnz], INT32 — CSC row indices (forward output[1])</li>
 *   <li>{@code cscColPtr}     – 1D [cols+1], INT32 — CSC column pointers (forward output[2])</li>
 *   <li>{@code gradCscValues} – 1D [nnz], float — upstream gradient w.r.t. cscValues output</li>
 * </ol>
 *
 * <p><b>Integer arguments (IArgs):</b> {@code rows}, {@code cols}
 *
 * <p><b>Output:</b>
 * <ol>
 *   <li>{@code dAValues} – 1D [nnz], same float dtype as {@code gradCscValues} —
 *       gradient w.r.t. the CSR values input</li>
 * </ol>
 *
 * <p>This op is forward-only (no {@code doDiff}); it is the gradient leaf for
 * {@link CsrToCsc#doDiff}.
 */
public class CsrToCscBp extends DynamicCustomOp {

    /** No-arg constructor required for ImportClassMapping reflection. */
    public CsrToCscBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param csrColIdx     1D [nnz] INT32/INT64 — CSR column indices (forward input[1])
     * @param csrRowPtr     1D [rows+1] INT32/INT64 — CSR row pointers (forward input[2])
     * @param cscRowIdx     1D [nnz] INT32 — CSC row indices (forward output[1])
     * @param cscColPtr     1D [cols+1] INT32 — CSC column pointers (forward output[2])
     * @param gradCscValues 1D [nnz] float — upstream gradient w.r.t. cscValues output
     * @param rows          number of rows in the original matrix A
     * @param cols          number of columns in the original matrix A
     */
    public CsrToCscBp(INDArray csrColIdx, INDArray csrRowPtr,
                       INDArray cscRowIdx, INDArray cscColPtr,
                       INDArray gradCscValues, long rows, long cols) {
        super(new INDArray[]{csrColIdx, csrRowPtr, cscRowIdx, cscColPtr, gradCscValues}, null);
        addIArgument(rows, cols);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd            SameDiff instance
     * @param csrColIdx     SDVariable [nnz] INT32/INT64 — CSR column indices
     * @param csrRowPtr     SDVariable [rows+1] INT32/INT64 — CSR row pointers
     * @param cscRowIdx     SDVariable [nnz] INT32 — CSC row indices
     * @param cscColPtr     SDVariable [cols+1] INT32 — CSC column pointers
     * @param gradCscValues SDVariable [nnz] float — upstream gradient w.r.t. cscValues
     * @param rows          number of rows
     * @param cols          number of columns
     */
    public CsrToCscBp(SameDiff sd,
                       SDVariable csrColIdx, SDVariable csrRowPtr,
                       SDVariable cscRowIdx, SDVariable cscColPtr,
                       SDVariable gradCscValues, long rows, long cols) {
        super(sd, new SDVariable[]{csrColIdx, csrRowPtr, cscRowIdx, cscColPtr, gradCscValues});
        addIArgument(rows, cols);
    }

    @Override
    public String opName() {
        return "csr_to_csc_bp";
    }

    /**
     * Output data type: same float dtype as {@code gradCscValues} (input index 4).
     */
    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Collections.singletonList(dataTypes.get(4));
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/csr_to_csc_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// csr_to_csc_bp — backward pass for csr_to_csc.
//
// Forward:  (csrValues[nnz], csrColIdx[nnz], csrRowPtr[rows+1]), IArgs (rows, cols)
//              → cscValues[nnz], cscRowIdx[nnz], cscColPtr[cols+1]
//
// The forward permutation P: CSR(A) → CSC(A) is computed by a counting-sort
// transpose.  The backward pass applies P⁻¹ via a second csr_to_csc call:
// treat (gradCscValues, cscRowIdx, cscColPtr) as the CSR of A^T (shape [cols, rows])
// and transpose back with dimensions (cols, rows) → values land in CSR(A) row-major order.
//
// Inputs:
//   [0] csrColIdx     [nnz],     INT32/INT64 — CSR column indices (forward input[1])
//   [1] csrRowPtr     [rows+1],  same INT — CSR row pointers (forward input[2])
//   [2] cscRowIdx     [nnz],     INT32 — CSC row indices (forward output[1])
//   [3] cscColPtr     [cols+1],  INT32 — CSC column pointers (forward output[2])
//   [4] gradCscValues [nnz],     float — upstream gradient w.r.t. cscValues output
// IArgs:
//   [0] rows
//   [1] cols
// Output:
//   [0] dAValues [nnz], same float dtype as gradCscValues — gradient w.r.t. csrValues input
//
// csrColIdx and csrRowPtr are accepted for structural validation but are not
// used in the computation (the helper only needs the CSC arrays to invert P).
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_csr_to_csc_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_convert_bp.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(csr_to_csc_bp, 5, 1, false, 0, 2) {
  auto csrColIdx     = INPUT_VARIABLE(0);
  auto csrRowPtr     = INPUT_VARIABLE(1);
  auto cscRowIdx     = INPUT_VARIABLE(2);
  auto cscColPtr     = INPUT_VARIABLE(3);
  auto gradCscValues = INPUT_VARIABLE(4);

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);
  const LongType nnz  = gradCscValues->lengthOf();

  // ── Validation ─────────────────────────────────────────────────────────────
  REQUIRE_TRUE(csrColIdx->rankOf() == 1, 0,
               "csr_to_csc_bp: csrColIdx must be 1D, got rank %d", csrColIdx->rankOf());
  REQUIRE_TRUE(csrRowPtr->rankOf() == 1, 0,
               "csr_to_csc_bp: csrRowPtr must be 1D, got rank %d", csrRowPtr->rankOf());
  REQUIRE_TRUE(cscRowIdx->rankOf() == 1, 0,
               "csr_to_csc_bp: cscRowIdx must be 1D, got rank %d", cscRowIdx->rankOf());
  REQUIRE_TRUE(cscColPtr->rankOf() == 1, 0,
               "csr_to_csc_bp: cscColPtr must be 1D, got rank %d", cscColPtr->rankOf());
  REQUIRE_TRUE(gradCscValues->rankOf() == 1, 0,
               "csr_to_csc_bp: gradCscValues must be 1D, got rank %d", gradCscValues->rankOf());
  REQUIRE_TRUE(csrRowPtr->lengthOf() == rows + 1, 0,
               "csr_to_csc_bp: csrRowPtr length must be rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)csrRowPtr->lengthOf());
  REQUIRE_TRUE(cscColPtr->lengthOf() == cols + 1, 0,
               "csr_to_csc_bp: cscColPtr length must be cols+1=%lld, got %lld",
               (long long)(cols + 1), (long long)cscColPtr->lengthOf());
  REQUIRE_TRUE(csrColIdx->lengthOf() == nnz, 0,
               "csr_to_csc_bp: csrColIdx length must match gradCscValues length (%lld), got %lld",
               (long long)nnz, (long long)csrColIdx->lengthOf());
  REQUIRE_TRUE(cscRowIdx->lengthOf() == nnz, 0,
               "csr_to_csc_bp: cscRowIdx length must match gradCscValues length (%lld), got %lld",
               (long long)nnz, (long long)cscRowIdx->lengthOf());

  // ── Execute ────────────────────────────────────────────────────────────────
  auto dAValues = OUTPUT_VARIABLE(0);

  if (nnz > 0) {
    sd::ops::helpers::csr_to_csc_bp(*cscRowIdx, *cscColPtr, *gradCscValues, *dAValues, rows, cols);
  }

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(csr_to_csc_bp) {
  // dAValues has same shape as csrColIdx (nnz elements) and dtype of gradCscValues.
  auto csrColIdx     = INPUT_VARIABLE(0);
  auto gradCscValues = INPUT_VARIABLE(4);

  const LongType nnz = csrColIdx->lengthOf();
  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(
      gradCscValues->dataType(), 'c', {nnz});
  return SHAPELIST(outShape);
}

DECLARE_TYPES(csr_to_csc_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})     // csrColIdx
      ->setAllowedInputTypes(1, {ALL_INTS})     // csrRowPtr
      ->setAllowedInputTypes(2, {ALL_INTS})     // cscRowIdx
      ->setAllowedInputTypes(3, {ALL_INTS})     // cscColPtr
      ->setAllowedInputTypes(4, {ALL_FLOATS})   // gradCscValues
      ->setAllowedOutputTypes(0, {ALL_FLOATS}); // dAValues
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_csr_to_csc_bp)
